### PR TITLE
Audio atoms

### PR DESCRIPTION
**Pull request by webb04**
Support audio atoms

## Screenshots

| src-foundations 2.4.0 | src-foundations 1.5.0 |
| --- | --- |
|  |  |

### CODE REVIEW

File: src/server/csp.ts [hunk 1]
```diff
@@ -67,12 +67,13 @@ const assetHashes = (assets: string[]): string =>
 
 const buildCsp = ({ styles, scripts }: Assets, twitter: boolean): string => `
     default-src 'self';
-    style-src 'self' ${assetHashes(styles)} https://interactive.guim.co.uk ${twitter ? 'https://platform.twitter.com' : ''};
+    style-src 'self' 'unsafe-inline';
```
Comment (JamieB-gu) [inline src/server/csp.ts:70]: I see you've flagged this in the description already, but just a reminder that we probably want to change this before merging.

File: src/atoms.ts [hunk 2]
```diff
@@ -246,6 +246,28 @@ function parseAtom(
...
+            const atom = atoms.audios?.find(audio => audio.id === id);
+
+            if (atom?.data?.kind !== "audio") {
+                return err(`No atom matched this id: ${id}`);
+            }
+            const { id: audioId, title } = atom;
+            const { kicker, trackUrl } = atom?.data?.audio;
+
+            if (!title) {
+                return err(`No videoId for atom: ${id}`);
```
Comment (JamieB-gu) [inline src/atoms.ts:259]:
```suggestion
                return err(`No title for audio atom with id: ${audioId}`);
```

File: src/client/article.ts [hunk 3]
```diff
@@ -222,6 +224,21 @@ function hasSeenCards(): void {
...
+    Array.from(document.querySelectorAll('.js-audio-atom'))
+        .forEach(atom => {
+            const id = atom.getAttribute('id');
+            const trackUrl = atom.getAttribute('trackurl');
+            const kicker = atom.getAttribute('kicker');
+            const title = atom.getAttribute('title');
+            const pillar = Pillar.Culture;
+
+            if (id && trackUrl && kicker && title && pillar) {
+                ReactDOM.hydrate(h(AudioAtom, { id, trackUrl, kicker, title, pillar }), atom)
```
Comment (JamieB-gu) [inline src/client/article.ts:237]: Is there any way we could render audio atoms server-side @shtukas ?

File: src/renderer.ts [hunk 4]
```diff
@@ -595,6 +595,11 @@ const render = (format: Format, excludeStyles = false) =>
             const figcaption = h(FigCaption, { format, caption, credit: none });
             return styledH('figure', figureAttributes, [ styledH('div', attributes), figcaption ]);
         }
+
+        case ElementKind.AudioAtom: {
+            const { pillar } = format;
+            return h('div', { ...element, pillar: pillar.valueOf(), className: 'js-audio-atom audio-atom' }, h(AudioAtom, { ...element, pillar }))
```
Comment (JamieB-gu) [inline src/renderer.ts:601]: If `...element` and `pillar` exist to add attributes to the `<div>` could we use [`data-*`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/data-*) attributes instead?

What do `valueOf` and the `audio-atom` class do here?

File: src/server/page.tsx [hunk 5]
```diff
@@ -71,6 +71,10 @@ const styles = (format: Format): string => `
         line-height: 1.5;
     }
 
+    .audio-atom figure {
+        margin: 0;
+    }
```
Comment (JamieB-gu) [inline src/server/page.tsx:76]: Can we apply this directly in the `renderer` using Emotion?

**Review by JamieB-gu — COMMENTED** [review]
Looks good, a few comments/questions.

File: src/renderer.ts (hunk 4 above)
Comment (webb04) [reply src/renderer.ts:601]: Passing `Pillar.Sport` doesn't seem to work as no pillar styles are added (The screenshot above is missing the blue sport colours).

`.valueOf` isn't working either actually, I need to take another look at this.

Comment (webb04) [reply src/renderer.ts:601]: Had to downgrade our src-foundations library to match dotcom in order to see the pillar colours`"@guardian/src-foundations": "^1.5.0"`

cc @shtukas

File: package.json [hunk 6]
```diff
@@ -42,12 +42,12 @@
     "@guardian/src-brand": "^2.4.0",
     "@guardian/src-button": "^2.4.0",
     "@guardian/src-ed-lines": "^2.4.0",
-    "@guardian/src-foundations": "^2.4.0",
+    "@guardian/src-foundations": "^1.5.0",
     "@guardian/src-icons": "^2.4.0",
-    "@guardian/src-radio": "^2.4.0",
+    "@guardian/src-radio": "^1.5.0",
```
Comment (JamieB-gu) [inline package.json:47]: Would it be better to fix this upstream in `atoms-rendering` rather than downgrade our version of Source?

Comment (webb04) [reply package.json:47]: That could be possible. I also had to downgrade src-foundations to 2.0.0 on the discussion-rendering branch to be compatible, so there might be more work involved in updating a few shared libraries

File: src/renderer.ts (hunk 4 above)
Comment (webb04) [reply src/renderer.ts]: Upgraded to atoms-rendering 2.0.0 which fixes the problem

File: src/renderer.ts [hunk 7]
```diff
@@ -628,6 +628,17 @@ const render = (format: Format, excludeStyles = false) =>
                 });
                 return styledH('figure', figureAttributes, [ styledH('div', attributes), figcaption ]);
             }
+
+            case ElementKind.AudioAtom: {
+                const { theme } = format;
+                const pillar = themeFromString('pillar/' + themeToPillar(theme));
```
Comment (JamieB-gu) [inline src/renderer.ts:634]: I'm a little puzzled by this line... there seem to be a couple of conversions back and forth?

**Review by JamieB-gu — APPROVED** [review]
Looks good!

**Pull request merged**